### PR TITLE
Support __half in histogram

### PR DESCRIPTION
**Pull request by senior-zero**
This PR addresses the following [issue](https://github.com/NVIDIA/cub/issues/484). Since there's another histogram related PR, I've tried to keep changes to the dispatch layer minimalistic.

### CODE REVIEW

File: cub/device/dispatch/dispatch_histogram.cuh [hunk 1]
```diff
@@ -175,6 +175,26 @@ template <
...
+      return static_cast<T>(upper_level - lower_level) / bins;
+    }
+
+#if defined(__CUDA_FP16_TYPES_EXIST__)
+    // There are no host versions of arithmetic operations on `__half`, so 
+    // all arithmetic operations on host shall be done on `float`
+    CUB_RUNTIME_FUNCTION 
+    static __half ComputeScale(__half lower_level, __half upper_level, int bins)
+    {
+      return __float2half(
```
Comment (allisonvacanti) [inline cub/device/dispatch/dispatch_histogram.cuh:192]: Question: Should this also check for SM53, or are the conversion intrinsics always necessary here?

File: test/test_device_histogram.cu [hunk 2]
```diff
@@ -44,8 +44,14 @@
 
 #include "test_util.h"
 
-using namespace cub;
+#define TEST_HALF_T \
+  (__CUDACC_VER_MAJOR__ >= 9 || CUDA_VERSION >= 9000) && !_NVHPC_CUDA
```
Comment (allisonvacanti) [inline test/test_device_histogram.cu:48]: Question: Why `!_NVHPC_CUDA`?

File: test/test_device_histogram.cu [hunk 3]
```diff
@@ -1399,6 +1593,10 @@ int main(int argc, char** argv)
 
     TestLevelsAliasing();
 
+#if (__CUDACC_VER_MAJOR__ >= 9 || CUDA_VERSION >= 9000) && !_NVHPC_CUDA
```
Comment (allisonvacanti) [inline test/test_device_histogram.cu:1596]:
```suggestion
#if TEST_HALF_T
```

File: test/test_device_histogram.cu [hunk 4]
```diff
@@ -256,6 +388,48 @@ struct Dispatch<1, 1, CUB>
         return error;
     }
 
+#if defined(TEST_HALF_T)
```
Comment (allisonvacanti) [inline test/test_device_histogram.cu:391]:
```suggestion
#if TEST_HALF_T
```

File: test/test_device_histogram.cu [hunk 5]
```diff
@@ -212,6 +304,46 @@ struct Dispatch<1, 1, CUB>
...
```
Comment (allisonvacanti) [inline test/test_device_histogram.cu:307]:
```suggestion
#if TEST_HALF_T
```

File: test/test_device_histogram.cu [hunk 6]
```diff
@@ -164,6 +212,50 @@ struct Dispatch<NUM_ACTIVE_CHANNELS, NUM_CHANNELS, CUB>
...
```
Comment (allisonvacanti) [inline test/test_device_histogram.cu:215]:
```suggestion
#if TEST_HALF_T
```

File: test/test_device_histogram.cu [hunk 7]
```diff
@@ -118,6 +124,50 @@ struct Dispatch<NUM_ACTIVE_CHANNELS, NUM_CHANNELS, CUB>
...
```
Comment (allisonvacanti) [inline test/test_device_histogram.cu:127]:
```suggestion
#if TEST_HALF_T
```

**Review by allisonvacanti — CHANGES_REQUESTED** [review]

File: cub/device/dispatch/dispatch_histogram.cuh (hunk 1 above)
Comment (senior-zero) [reply cub/device/dispatch/dispatch_histogram.cuh:192]: SM53 introduces arithmetic operations. Conversions are [present](https://godbolt.org/z/57xdqsdev) in SM35.

File: test/test_device_histogram.cu (hunk 2 above)
Comment (senior-zero) [reply test/test_device_histogram.cu:48]: > ────────────────────────────────────────────────┐
    commit e882cfd875417589276586f4657f285885c523b9 │
   ────────────────────────────────────────────────┘
    Author: David Olsen <dolsen@nvidia.com>
    Date:   Thu Nov 5 14:39:01 2020 -0800
    Fix cub/util_type.cuh to not use 16-bit float with NVC++
    A recent commit changed several occurances of "__CUDACC_VER_MAJOR__ >= 9"
    in preprocessor expressions to "(__CUDACC_VER_MAJOR__ >= 9 ||
    CUDA_VERSION >= 9000)".  NVC++ defines CUDA_VERSION but not
    __CUDACC_VER_MAJOR__, so the commit changed the condition from false to
    true for NVC++.  The code inside those blocks doesn't work with NVC++,
    causing compilation errors.  Fix the regression by changing the expressions
    to "(__CUDACC_VER_MAJOR__ >= 9 || CUDA_VERSION >= 9000) &&
    !__NVCOMPILER_CUDA__".

Comment (allisonvacanti) [inline test/test_device_histogram.cu:48]:
```suggestion
  (__CUDACC_VER_MAJOR__ >= 9 || CUDA_VERSION >= 9000) && !defined(_NVHPC_CUDA)
```

**Review by allisonvacanti — APPROVED** [review]

**Comment by allisonvacanti** [comment]
Approved pending minor fixes and retesting.

**Pull request merged**